File: java_pro/src/main/java/jv/com/tencent/cain/design_pattern/fly_weight/FlyWeightFactory.java
package jv.com.tencent.cain.design_pattern.fly_weight;

import java.util.HashMap;
import java.util.Locale;

/**
 * @author cainjiang
 * @date 2020/8/14
 */
public class FlyWeightFactory {
    private final HashMap<String, AbstractFlyWeight> flyWeights = new HashMap<>();
    private final AbstractFlyWeight unSharedFlyWeight = new UnSharedFlyWeight();

    public AbstractFlyWeight getUnSharedFlyWeight() {
        return unSharedFlyWeight;
    }

    public FlyWeightFactory() {
        flyWeights.put("X", new ConcretFlyWeight());
        flyWeights.put("Y", new ConcretFlyWeight());
        flyWeights.put("Z", new ConcretFlyWeight());
    }

    public AbstractFlyWeight getFlyWeight(String key) {
        key = key.toUpperCase(Locale.ENGLISH);
        AbstractFlyWeight flyWeight;
        flyWeight = flyWeights.get(key);
        return flyWeight;
    }

    int getFlyWeightCount() {
        return flyWeights.size() + 1;
    }

}

File: java_pro/src/main/kotlin/kt/com/tencent/cain/desigin_pattern/fly_weight/FlyWeightFactory.kt
package kt.com.tencent.cain.desigin_pattern.fly_weight

import java.util.*

/**
 * @author cainjiang
 * @date 2020/8/14
 */
class FlyWeightFactory {
    private val flyWeights by lazy {
        hashMapOf<String, BaseFlyWeight>()
    }

    val unSharedFlyWeight by lazy {
        UnSharedFlyWeight()
    }

    init {
        flyWeights["A"] = FlyWeight()
        flyWeights["B"] = FlyWeight()
        flyWeights["C"] = FlyWeight()
    }

    fun getFlyWeightByName(key: String): BaseFlyWeight? {
        return flyWeights[key.toUpperCase(Locale.ENGLISH)]
    }

    fun getFlyWeightCount(): Int {
        return flyWeights.size + 1
    }

}